src/DFA/DFA.fs

```diff
@@ -0,0 +1,210 @@
+            let mutable newStartState = Unchecked.defaultof<_>
+            let toProcess = Stack<State<_, _> * (State<_, _> -> unit)>[(state, (fun _ -> ()))]
+            let copyOfState = Dictionary<_, _>()
+
+            while toProcess.Count > 0 do
+                let state, addTransitions = toProcess.Pop()
+                let exists, newState = copyOfState.TryGetValue state
+
+                if not exists then
+                    let newState = State<_, _>(state.Id, state.IsStart, not state.IsFinal)
+
+                    if newState.IsStart then
+                        newStartState <- newState
+
+                    addTransitions newState
+
+                    for (char, targetState) in state.GetAllTransitions() do
+                        toProcess.Push(targetState, (fun newTargetState -> newState.AddTransition char newTargetState))
+
+                    copyOfState.Add(state, newState)
+                else
+                    addTransitions newState
+
+            newStartState
+
+        let newStartState = dfs dfa.StartState
+        DFA(newStartState)
+
+    let intersection (dfa1: DFA<_, _>) (dfa2: DFA<_, _>) =
+        let dfs (state1: State<_, _>) (state2: State<_, _>) =
+            let mutable newStartState = Unchecked.defaultof<_>
+
+            let toProcess =
+                Stack<State<_, _> * State<_, _> * (State<_, _> -> unit)>[(state1, state2, (fun _ -> ()))]
+
+            let copyOfState = Dictionary<_, _>()
+
+            while toProcess.Count > 0 do
+                let state1, state2, addTransitions = toProcess.Pop()
+                let exists, newState = copyOfState.TryGetValue((state1, state2))
+
+                if not exists then
+                    let newState =
+                        State<_, _>(
+                            $"{state1.Id}_{state2.Id}",
+                            state1.IsStart && state2.IsStart,
+                            state1.IsFinal && state2.IsFinal
+                        )
+
+                    if newState.IsStart then
+                        newStartState <- newState
+
+                    addTransitions newState
+
+                    for (char1, targetState1) in state1.GetAllTransitions() do
+                        for (char2, targetState2) in state2.GetAllTransitions() do
+                            if char1 = char2 then
+                                toProcess.Push(
+                                    targetState1,
+                                    targetState2,
+                                    (fun newTargetState -> newState.AddTransition char1 newTargetState)
+                                )
+
+                    copyOfState.Add((state1, state2), newState)
+                else
+                    addTransitions newState
+
+            newStartState
+
+        let newStartState = dfs dfa1.StartState dfa2.StartState
+        DFA(newStartState)
+
+    let union dfa1 dfa2 =
+        complement (intersection (complement dfa1) (complement dfa2))
```

Comment:
In my opinion it is expensive to do three complements and intersection. I think the better way to do this is to have generic DFS function for union and intersection with a way to pass custom State constructor.

Reply:
Added as task to #3. Accepting now